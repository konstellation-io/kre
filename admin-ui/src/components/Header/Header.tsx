import { useQuery } from '@apollo/react-hooks';
import { loader } from 'graphql.macro';
import { get } from 'lodash';
import React, { FunctionComponent, ReactElement } from 'react';
import Settings from '../../components/Settings/Settings';
import { GetMe } from '../../graphql/queries/types/GetMe';
import styles from './Header.module.scss';

const GetMeQuery = loader('../../graphql/queries/getMe.graphql');

type Props = {
  children?: ReactElement | ReactElement[] | null;
  hideSettings?: boolean;
};
const Header: FunctionComponent<Props> = ({
  children,
  hideSettings = false
}) => {
  const { data, loading } = useQuery<GetMe>(GetMeQuery);

  if (loading)
    return <div className={styles.splash} data-testid="splashscreen" />;

  const username: string = get(data, 'me.email');

  return (
    <header className={styles.container} data-testid="app-header">
      <img
        className={styles.konstellationText}
        src={'/img/brand/konstellation.svg'}
        alt="konstellation"
      />
      <div className={styles.customHeaderElements}>{children}</div>
      {!hideSettings && <Settings label={username} />}
    </header>
  );
};

export default Header;
