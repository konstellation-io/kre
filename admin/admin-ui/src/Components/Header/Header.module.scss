@import 'Styles/colors';
@import 'Styles/mixins';
@import 'Styles/variables';

.container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex-shrink: 0;
  padding: 0 8px 0 13px;
  width: 100%;
  height: $header-height;
  background-color: $bg-color-dark;
  white-space: nowrap;
  z-index: $z-index-header;
}

.konstellationIcon {
  height: 21px;
  margin: 0 7px 0 9px;
}
.konstellationText {
  height: 30px;
  margin-right: 20px;
}

.customHeaderElements {
  height: 100%;
  display: flex;
  flex-direction: row-reverse;
  flex-grow: 1;
  align-items: center;

  flex-wrap: wrap;
  overflow: hidden;
}

.splash {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: $bg-color-dark;
  z-index: 100;
}
