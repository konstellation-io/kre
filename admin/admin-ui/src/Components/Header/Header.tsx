import React, { FunctionComponent, ReactElement, memo } from 'react';

import { GetMe } from 'Graphql/queries/types/GetMe';
import Settings from 'Components/Settings/Settings';
import { get } from 'lodash';
import { loader } from 'graphql.macro';
import styles from './Header.module.scss';
import { useQuery } from '@apollo/react-hooks';

const GetMeQuery = loader('Graphql/queries/getMe.graphql');

type Props = {
  children?: ReactElement | ReactElement[] | null;
  hideSettings?: boolean;
};
const Header: FunctionComponent<Props> = ({
  children,
  hideSettings = false
}) => {
  const { data, loading } = useQuery<GetMe>(GetMeQuery);

  if (loading)
    return <div className={styles.splash} data-testid="splashscreen" />;

  const username = get(data?.me, 'email', '');

  return (
    <header className={styles.container} data-testid="app-header">
      <img
        className={styles.konstellationText}
        src={'/img/brand/konstellation.svg'}
        alt="konstellation"
      />
      <div className={styles.customHeaderElements}>{children}</div>
      {!hideSettings && <Settings label={username} />}
    </header>
  );
};

export default memo(Header);
