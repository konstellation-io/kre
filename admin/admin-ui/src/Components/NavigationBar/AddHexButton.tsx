import React from 'react';

import styles from './NavigationBar.module.scss';

function AddHexButton() {
  return (
    <div className={styles.addButton}>
      <div className={styles.buttonWrapper}>
        <div className={styles.buttonContainer}>
          <div className={styles.button} />
        </div>
      </div>
      <div className={styles.buttonWrapper}>
        <div className={styles.buttonContainer}>
          <div className={styles.button} />
          <div className={styles.buttonText}>{'+'}</div>
        </div>
      </div>
    </div>
  );
}

export default AddHexButton;
