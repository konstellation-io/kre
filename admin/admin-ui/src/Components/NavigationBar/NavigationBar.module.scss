@import 'Styles/colors';
@import 'Styles/mixins';

$color-hexagon: palette(base, 400);
$color-hexagon-active: $color-primary;
$color-hexagon-hover: palette(highlight, 700);
$color-hexagon-disabled: palette(base, 200);

.container {
  padding-top: 16px;
  z-index: 2;

  .link:last-child:after {
    background-color: transparent;
  }
}

.link {
  position: relative;

  &:after {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 4px;
    height: 40px;
    background-color: $color-hexagon;
    transition: background-color ease 0.4s;
  }
}

.buttonWrapper {
  height: 40px;
  width: 40px;
  margin-bottom: 24px;

  &:hover {
    .button {
      background-color: $color-hexagon-hover;
      box-shadow: -26px -2px 29px -2px rgba($color-hexagon-active, 0.3);
    }
    .buttonText {
      color: font-color(dark);
    }
  }
  &.disabled {
    .button {
      background-color: $color-hexagon-disabled;
      cursor: not-allowed;
    }
    .buttonText {
      color: font-color(dark);
    }
    &:hover {
      .button {
        box-shadow: unset;
      }
    }
  }
}

.buttonContainer {
  @include hexagonWrapper;

  position: absolute;
  left: 15px;

  width: 40px;
  height: 40px;
}

.button {
  @include font-button;
  color: font-color(regular);

  @include hexagon;

  background-color: $color-hexagon;
  transition: background-color ease 0.2s, box-shadow ease 0.2s;
}

.active {
  &:after {
    background-color: $color-hexagon-active;
  }

  .button {
    background-color: $color-hexagon-active;
    box-shadow: -26px -2px 29px -2px rgba($color-hexagon-active, 0.4);
  }

  .buttonText {
    color: font-color(dark);
  }
}

.buttonText {
  @include center;

  @include font-button;
  color: font-color(light);

  transition: color ease 0.3s;

  pointer-events: none;
}

.miniHex {
  .buttonWrapper,
  .buttonContainer {
    height: 14px;
    width: 14px;
  }
}

.buttonWrapper {
  .miniHex:nth-child(1) .buttonContainer {
    left: 17px;
    top: 0px;
  }
  .miniHex:nth-child(2) .buttonContainer {
    left: 31px;
    top: 8px;
  }
  .miniHex:nth-child(3) .buttonContainer {
    left: 17px;
    top: 16px;
  }
  .miniHex:nth-child(4) .buttonContainer {
    left: 31px;
    top: 24px;
  }
}

.addButton {
  .buttonWrapper:first-child {
    pointer-events: none;
  }

  .buttonWrapper:last-child {
    position: absolute;
    top: 2px;
    left: 2px;

    &,
    .buttonContainer {
      height: 36px;
      width: 36px;
    }

    .button {
      background-color: $bg-color-dark;
    }

    &:hover {
      .button {
        background-color: #38a7c1;
      }
    }

    .buttonText {
      font-size: 16px;
    }
  }
}
