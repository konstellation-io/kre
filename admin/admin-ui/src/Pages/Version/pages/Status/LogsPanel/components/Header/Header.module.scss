@import 'Styles/mixins';
@import 'Styles/funcs';
@import 'Styles/shadows';
@import 'Styles/colors';

.container {
  display: flex;
  align-items: center;
  cursor: pointer;
  transition: height ease 0.3s, box-shadow linear 0.2s;
  box-shadow: 0;
  min-height: 3 * $grid-unit;

  &.opened {
    height: 4 * $grid-unit;
    flex-shrink: 0;
    pointer-events: none;

    .buttons {
      opacity: 1;
    }
  }

  &:not(.opened):hover {
    @include glow(4);

    .title {
      color: font-color(light);
    }
  }
}

.title {
  @include font-caption;
  color: font-color(lowlight);

  display: flex;
  align-items: center;
  margin: 0 20px 0 20px;
  transition: color ease 0.3s;

  svg {
    margin: 4px 1.5 * $grid-unit 0 0;
  }
}

.runtime {
  @include font-body;
  color: font-color(light);
}

.version {
  @include font-body;
  color: font-color(light);

  margin-left: 16px;
}

.buttons {
  display: flex;
  flex-direction: row;
  transition: opacity ease 0.4s;
  margin: 4px 8px 0 auto;
  pointer-events: all;

  div {
    margin-left: 8px;
    cursor: pointer;

    &:hover {
      color: font-color(light);
    }
  }
}

.stickBottom {
  transition: color ease 0.4s, background-color ease 0.4s;

  &.active {
    color: font-color(light);
    background-color: $bg-color-lighter;

    &:hover {
      color: font-color(light);
    }
  }
}
