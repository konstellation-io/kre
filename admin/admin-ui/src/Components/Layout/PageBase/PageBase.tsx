import React, { ReactElement, useEffect, useState } from 'react';

import { GET_LOG_TABS } from 'Graphql/client/queries/getLogs.graphql';
import Header from '../../Header/Header';
import NavigationBar from '../../NavigationBar/NavigationBar';
import cx from 'classnames';
import styles from './PageBase.module.scss';
import { useQuery } from '@apollo/react-hooks';

type PageBaseProps = {
  children: ReactElement | ReactElement[] | null;
  headerChildren?: ReactElement | ReactElement[] | null;
  customClassname?: string;
};

function PageBase({
  children,
  headerChildren,
  customClassname
}: PageBaseProps) {
  const [opened, setOpened] = useState<boolean>(false);

  const { data: localData } = useQuery(GET_LOG_TABS);

  useEffect(() => {
    setOpened(localData?.logTabs?.length);
  }, [localData]);
  return (
    <>
      <Header>{headerChildren}</Header>
      <div
        className={cx(styles.content, customClassname, {
          [styles.withLogs]: opened
        })}
      >
        <NavigationBar />
        {children}
      </div>
    </>
  );
}

export default PageBase;
