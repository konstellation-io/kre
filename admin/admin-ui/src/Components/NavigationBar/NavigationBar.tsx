import { ErrorMessage, SpinnerCircular, VerticalBar } from 'kwc';
import {
  GetRuntimes,
  GetRuntimes_runtimes
} from 'Graphql/queries/types/GetRuntimes';
import React, { memo } from 'react';

import AddHexButton from './AddHexButton';
import ConditionalLink from '../ConditionalLink/ConditionalLink';
import HexButton from './HexButton';
import { MONORUNTIME_MODE } from 'index';
import MultiHexButton from './MultiHexButton';
import { NavLink } from 'react-router-dom';
import ROUTE from 'Constants/routes';
import { RuntimeStatus } from 'Graphql/types/globalTypes';
import { buildRoute } from 'Utils/routes';
import { checkPermission } from 'rbac-rules';
import cx from 'classnames';
import { get } from 'lodash';
import { loader } from 'graphql.macro';
import styles from './NavigationBar.module.scss';
import { useQuery } from '@apollo/client';
import useUserAccess from 'Hooks/useUserAccess';

const GetRuntimesQuery = loader('Graphql/queries/getRuntimes.graphql');

function NavigationBar() {
  const { accessLevel } = useUserAccess();
  const { data, loading, error } = useQuery<GetRuntimes>(GetRuntimesQuery);

  if (loading) return <SpinnerCircular />;
  if (error) return <ErrorMessage />;

  const runtimes: GetRuntimes_runtimes[] = get(data, 'runtimes', []);

  const buttons = runtimes.map((runtime: GetRuntimes_runtimes) => {
    const disabled = runtime.status === RuntimeStatus.CREATING;
    return (
      <ConditionalLink
        key={runtime.name}
        LinkType={NavLink}
        to={buildRoute.runtime(ROUTE.RUNTIME, runtime.id)}
        disabled={disabled}
        linkProps={{
          activeClassName: styles.active,
          className: styles.link
        }}
      >
        <HexButton label={runtime.name} disabled={disabled} />
      </ConditionalLink>
    );
  });

  const canAccessRuntimeList = !MONORUNTIME_MODE;
  if (canAccessRuntimeList) {
    buttons.unshift(
      <NavLink
        key={`NavBarItem_Runtimes`}
        to={ROUTE.HOME}
        activeClassName={styles.active}
        className={styles.link}
        exact={true}
      >
        <MultiHexButton />
      </NavLink>
    );
  }

  const canAddRuntime =
    checkPermission(accessLevel, 'runtime:edit') && !MONORUNTIME_MODE;
  if (canAddRuntime) {
    buttons.push(
      <NavLink
        key={`NavBarItem_AddRuntime`}
        to={ROUTE.NEW_RUNTIME}
        activeClassName={styles.active}
        className={cx(styles.link, styles.addRuntimeLink)}
      >
        <AddHexButton />
      </NavLink>
    );
  }

  return (
    <VerticalBar>
      <div className={styles.container} data-testid="navigation-bar">
        {buttons}
      </div>
    </VerticalBar>
  );
}

export default memo(NavigationBar);
