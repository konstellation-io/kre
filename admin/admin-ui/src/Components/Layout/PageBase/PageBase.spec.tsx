import Header from '../../Header/Header';
import NavigationBar from '../../NavigationBar/NavigationBar';
import PageBase from './PageBase';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('@apollo/client', () => ({
  useQuery: jest.fn(() => ({})),
  gql: jest.fn()
}));
jest.mock('Graphql/client/cache', () => ({}));

describe('PageBase', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(
      <PageBase>
        <div className="element">Element</div>
      </PageBase>
    );
  });

  it('matches snapshot', async () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('contains right components', async () => {
    expect(wrapper.exists(Header)).toBeTruthy();
    expect(wrapper.exists(NavigationBar)).toBeTruthy();
    expect(wrapper.exists('.element')).toBeTruthy();
  });

  it('allows the addition of children to the header', async () => {
    const headerChildren = <span className="headerChildren">Child</span>;
    wrapper.setProps({ headerChildren });

    expect(wrapper.find(Header).exists('.headerChildren')).toBeTruthy();
    expect(wrapper.find('.content').exists('.headerChildren')).toBeFalsy();
  });
});
