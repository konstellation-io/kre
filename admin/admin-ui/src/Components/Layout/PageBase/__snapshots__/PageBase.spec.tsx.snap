// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBase matches snapshot 1`] = `
<Fragment>
  <Memo(Header) />
  <div
    className="content"
  >
    <Memo(NavigationBar) />
    <div
      className="element"
    >
      Element
    </div>
  </div>
</Fragment>
`;
